hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientInvocationServiceImpl.java

```diff
@@ -287,29 +294,29 @@ public boolean isSmartRoutingEnabled() {
         return isSmartRoutingEnabled;
     }
 
-    private class CleanResourcesTask implements Runnable {
+    private class BackupTimeoutTask implements Runnable {
         @Override
         public void run() {
             for (ClientInvocation invocation : invocations.values()) {
-                ClientConnection connection = invocation.getSendConnection();
-                if (connection == null) {
-                    continue;
-                }
+                invocation.detectAndHandleBackupTimeout(operationBackupTimeoutMillis);
+            }
+        }
+    }
 
-                if (!connection.isAlive()) {
-                    notifyException(invocation, connection);
-                    continue;
-                }
+    private class ConnectionRemovedListener implements ConnectionListener<ClientConnection> {
+        @Override
+        public void connectionAdded(ClientConnection connection) {
 
-                if (isBackupAckToClientEnabled) {
-                    invocation.detectAndHandleBackupTimeout(operationBackupTimeoutMillis);
-                }
-            }
         }
 
-        private void notifyException(ClientInvocation invocation, ClientConnection connection) {
-            Exception ex = new TargetDisconnectedException(connection.getCloseReason(), connection.getCloseCause());
-            invocation.notifyException(ex);
+        @Override
+        public void connectionRemoved(ClientConnection connection) {
+            for (ClientInvocation invocation : invocations.values()) {
```

Comment:
Can this and the 
```
       if (!connection.write(clientMessage)) {
            if (invocation.getPermissionToNotifyForDeadConnection(connection)) {
                IOException exception = new IOException("Packet not sent to " + connection.getRemoteAddress() + " "
                        + clientMessage);
                invocation.notifyExceptionWithOwnedPermission(exception);
            }	            }
            return false;	        
       }
``` 

race?

I mean, lets say we registered the invocation and set the `sentConnection`, then the connection is closed, and we came here, acquire the permission, and retry the invocation and set the `sentConnection` to something alive. Then the code I posted above will not be able to do `SENT_CONNECTION.compareAndSet(this, deadConnection, null)` in `getPermissionToNotifyForDeadConnection` and spin

Reply:
The code is simplified, we don't have the while loop anymore.